### ISSUE TITLE
Next steps

### ISSUE THREAD

**Issue by b-remy** [JonnyyTorres/Galsim_JAX#1]
Hi @EiffL, with @JonnyyTorres we were thinking of what steps we should do next and made a small list:

- [x] Make a pip installable package
- [x] Train with forward model (PSF)
- [x] Compare our Resnet architecture against other implementations, e.g. 
  - [x] [flax example](https://github.com/google/flax/blob/main/examples/imagenet/models.py), [other flax code](https://github.com/fattorib/Flax-ResNets)
  - [x] stable diffusion's [autoencoder](https://github.com/Stability-AI/stablediffusion/blob/main/ldm/modules/diffusionmodules/model.py) and [resnet](https://github.com/Stability-AI/stablediffusion/blob/main/ldm/modules/diffusionmodules/model.py)
  - ([jax-lensing U-net](https://github.com/CosmoStat/jax-lensing/blob/master/jax_lensing/models/convdae.py))
- [ ] Add attention layers (like [stable diffusion's attention](https://github.com/Stability-AI/stablediffusion/blob/main/ldm/modules/attention.py))

Let us know your opinion and if you had anything else in mind :-)

**Comment by JonnyyTorres** [comment]
Hi @b-remy and @EiffL, I would like to mention that with the last [commit](https://github.com/JonnyyTorres/Galsim_JAX/commit/01fdeff26cf3f8d51584bec94d4b9f3851c53477) I made, the code is now in modular form and as a pip package as suggested by Benjamin, making the code more readable.

Additionally, the comparison and sample figures of the autoencoder are now linked as Artifacts each time a Wandb experiment is launched. Also, I added the regularizing term of the KL divergence as a FLAGS variable.

Another change is the addition of a FLAGS variable to indicate the GPU on which the training is performed. Here are some screenshots of the results I get in Wandb, where I can track the loss and log-likelihood, in addition to the figures mentioned above.  :-)

![loglog_loss](https://github.com/JonnyyTorres/Galsim_JAX/assets/92088528/13e46bb7-e862-4b54-964d-c954efba5a4e)
![loglog_log_likelihood](https://github.com/JonnyyTorres/Galsim_JAX/assets/92088528/ee3efc36-d8fd-4268-a0d9-7c8b52a97991)
![difference_pred](https://github.com/JonnyyTorres/Galsim_JAX/assets/92088528/2eb81f6e-866e-462b-88d4-7b54698a8b0a)
![samples_pred](https://github.com/JonnyyTorres/Galsim_JAX/assets/92088528/e92682cd-0da7-42e7-9ff7-08d866d35af9)

Finally, I would like to mention that yesterday I tried to run a training with 100000 epochs but apparently the gpu did not support.... do you have any idea of the possible cause?

**Comment by b-remy** [comment]
Thanks @JonnyyTorres, for the update!

I have a few comments:

On these lines, you might want to split the seed as well not to use the same one for the tow different sampling steps:
https://github.com/JonnyyTorres/Galsim_JAX/blob/4f543650740a75623e7ecede3b8b1d7e5bafa70f/VAE_Resnet_train.py#L369-L375
it seems that this is done correclty everywhere else.

On the latest plot, it is still auto-encoded galaxy images right? You don't sample points unconditionally in the latent space to generate images 

I don't think the GPUs cannot support running for a long training, do you have any log of your run?

**Comment by b-remy** [comment]
Nice :-)

Yes you might not want to go up to 100k steps for model comparison at this time. Is this log likelihood computed on the training set? 

I think that in order to properly compare models efficiency, we may want to look at the average log likelihood computed on the test set. If you saved the model weights, this should be a quick test to do.

**Comment by JonnyyTorres** [comment]
Yes, it is indeed the log likelihood for training. In the case of the test, I only do the log likelihood capture 10 times because it is computationally expensive, should I do it more times? 

Here is the link for the same metric for the test case: https://wandb.ai/jonnyytorres/galsim-jax-resnet/reports/test_log_likelihood-23-06-12-13-24-58---Vmlldzo0NjIxMjIx

**Comment by b-remy** [comment]
So, if I am not wrong, the two curves are the same model but one trained longer than the other. So looking at the [pink curve](https://wandb.ai/jonnyytorres/galsim-jax-resnet/reports/test_log_likelihood-23-06-12-13-24-58---Vmlldzo0NjIxMjIx), you don't have much improvement on the test set after 40k steps for this model.

**Comment by JonnyyTorres** [comment]
Yes, exactly. In that case, should I fix the steps to 40k - 50k maybe?

**Comment by b-remy** [comment]
Yep, I would go for it :-)

**Comment by b-remy** [comment]
Our next steps 

**To do**
- [ ] Full generative model with  Normalizing Flow in the latent space
- [ ] Compare statistics between training set & generated set (see [Fig. 4, 6, 7, 8, 9](https://arxiv.org/pdf/2008.03833.pdf))

**(To explore)**
- Latent diffusion model
	- train a (score-based) diffusion model in the latent space
	- use probability flow ODE to evaluate the likelihood

**Architecture development**
- [ ] softplus regularization (see if it cancels the noise pattern at the output of the decoder)
- [ ] Condition the encoder with the PSF

@JonnyyTorres , @EiffL
